File: 206 Reverse Linked List.c
/*
Reverse Linked List 
https://leetcode.com/problems/reverse-linked-list/

Iterative way & recursive way
*/

struct ListNode {
    int val;
    struct ListNode *next;
};
typedef struct ListNode lnode;

// recursively
lnode* reverseList(lnode* curr)
{
    lnode* head;
    if(curr == 0 || curr->next == 0) return curr;
    if(curr->next->next != 0)
        head = reverseList(curr->next);
    else
        head = curr->next;
    curr->next->next = curr;
    curr->next = 0;
    return head;
}
// iteratively
/*
lnode* reverseList(lnode* head)
{
    lnode *p0,*p1;
    if(head == 0) return 0;
    if(head->next == 0) return head;
    
    p0 = head->next;
    head->next = 0;
    while(p0->next != 0)
    {
        p1 = p0->next;
        p0->next = head;
        head = p0;
        p0 = p1;
    }
    p0->next = head;
    return p0;
}
*/

int main()
{
  return 0;
}

File: 206 Reverse Linked List.java
/*
206. Reverse Linked List
https://leetcode.com/problems/reverse-linked-list/

Reverse a singly linked list.

Example:

Input: 1->2->3->4->5->NULL
Output: 5->4->3->2->1->NULL
Follow up:

A linked list can be reversed either iteratively or recursively. Could you implement both?
*/
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    
    // Recursively
    public ListNode reverseList(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        
        ListNode newHead = dfs(head);
        head.next = null;
        return newHead;
    }
    
    /**
    *@node: != null
    *@return: new head of the reversed list
    */
    private ListNode dfs(ListNode node) {
        if (node.next == null) {
            return node;
        }
        
        ListNode newHead = dfs(node.next);
        
        // Redirect
        node.next.next = node;
        
        return newHead;
    }
}

class Solution {
    
    // Iteratively
    public ListNode reverseList(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        
        ListNode p1 = head.next;
        head.next = null;
        while (p1.next != null) {
            
            // Assign
            ListNode p2 = p1.next;
            
            // Redirect
            p1.next = head;
            
            // Move
            head = p1;
            p1 = p2;
        }
        p1.next = head;
        return p1;
    }
}
